### PR TITLE
[move-prover] Stackless bytecode refactoring.

### PR DESCRIPTION
**Pull request by wrwg**
This refactors stackless bytecode to make it better suitable for a program transformation approach of compilation to boogie.

- The number of opcodes has been drastically reduced by merging similar codes into one. For example, `Add(r, x, y)` becomes `Binary(Add, r, x, y)`. While a large flat enum of opcodes is good for the VM, it is not good for transformations which in many cases can deal with a class of basic ops generically (i.e. do something for all `Binary(op, r, x, y)`).
- Any dependencies from file_format in the bytecode have been removed. For example, instead of `SignatureToken` the bytecode uses now our own `Type`. Also, type instantiations are not longer represented by a signature pool index but directly.

On the way some incorrect treatment of generic type instantiations (likely introduced by file_format refactoring) has been fixed. There seem to be not much test coverage on prover side for this, so it didn't cause a problem until now.

Unfortunately, a set of unit tests for the stackless bytecode generator had to be removed. Those tests where still using old IR and have no way to depend on `spec-lang` which we now need. However, test coverage should be maintained because the translator is now exhaustively used by the prover itself.

## Motivation

Move boogie translation to transformational approach.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Existing tests pass. For some standard modules, verification has been turned off because it takes to long currently.

## Related PRs

NA

### CODE REVIEW

File: language/move-prover/stackless-bytecode-generator/src/stackless_bytecode.rs [hunk 1]
```diff
@@ -1,19 +1,81 @@
...
+/// A branch condition.
+#[derive(Debug, Clone, PartialEq, Eq, Copy)]
+pub enum BranchCond {
+    Always,
+    True(TempIndex),
+    False(TempIndex),
+}
+
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub enum StacklessBytecode {
```
Comment (shazqadeer) [inline language/move-prover/stackless-bytecode-generator/src/stackless_bytecode.rs:77]: We could rename StacklessBytecode -> Bytecode.

File: language/move-prover/stackless-bytecode-generator/src/stackless_bytecode_generator.rs [hunk 2]
```diff
@@ -16,53 +21,60 @@ use vm::{
...
-    pub local_types: Vec<SignatureToken>,
+    pub local_types: Vec<Type>,
     pub code: Vec<StacklessBytecode>,
 }
 
 pub struct StacklessBytecodeGenerator<'a> {
+    module_env: &'a ModuleEnv<'a>,
     module: &'a CompiledModule,
     function_definition_view: FunctionDefinitionView<'a, CompiledModule>,
     temp_count: usize,
```
Comment (shazqadeer) [inline language/move-prover/stackless-bytecode-generator/src/stackless_bytecode_generator.rs:32]: When I was reading the code yesterday, it looked to me as if we have the invariant temp_count == local_types.size.  I was tempted to remove temp_count.

File: language/move-prover/stackless-bytecode-generator/src/stackless_bytecode_generator.rs [hunk 3]
```diff
@@ -744,71 +834,79 @@ impl<'a> StacklessBytecodeGenerator<'a> {
...
                     *idx,
-                    None,
+                    vec![],
                 ));
             }
 
             Bytecode::MutBorrowGlobalGeneric(idx) | Bytecode::ImmBorrowGlobalGeneric(idx) => {
                 let struct_instantiation = self.module.struct_instantiation_at(*idx);
                 let struct_definition = self.module.struct_def_at(struct_instantiation.def);
+                let is_mut = matches!(bytecode, Bytecode::MutBorrowGlobalGeneric(..));
```
Comment (shazqadeer) [inline language/move-prover/stackless-bytecode-generator/src/stackless_bytecode_generator.rs:888]: Where is matches! defined?

File: language/move-prover/stackless-bytecode-generator/src/stackless_bytecode.rs (hunk 1 above)
Comment (wrwg) [reply language/move-prover/stackless-bytecode-generator/src/stackless_bytecode.rs:77]: Done

File: language/move-prover/stackless-bytecode-generator/src/stackless_bytecode_generator.rs (hunk 2 above)
Comment (wrwg) [reply language/move-prover/stackless-bytecode-generator/src/stackless_bytecode_generator.rs:32]: Not worth the effort I would say. There might be scenarios where the current thing is useful, even though apparently none in the code right now.

File: language/move-prover/stackless-bytecode-generator/src/stackless_bytecode_generator.rs (hunk 3 above)
Comment (wrwg) [reply language/move-prover/stackless-bytecode-generator/src/stackless_bytecode_generator.rs]: std::matches. Just came along it today for the first time.

**Pull request closed without merging**